File: app/src/main/java/com/mariotti/developer/futureclock/controllers/fragments/TimePickerFragment.java
package com.mariotti.developer.futureclock.controllers.fragments;

import android.app.Activity;
import android.app.Dialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.util.Log;
import android.widget.TimePicker;

public class TimePickerFragment extends DialogFragment
        implements TimePickerDialog.OnTimeSetListener {
    public static final String EXTRA_HOUR = "com.mariotti.developer.futureclock.controller.hour";
    public static final String EXTRA_MINUTE = "com.mariotti.developer.futureclock.controller.minute";

    private static final String TAG = "TimePickerFragment";
    private static final String ARG_HOUR = "ARG_HOUR";
    private static final String ARG_MINUTE = "ARG_MINUTE";

    public TimePickerFragment() {
    }

    public static TimePickerFragment newInstance(int hour, int minute) {
        TimePickerFragment fragment = new TimePickerFragment();
        Bundle args = new Bundle();
        args.putInt(ARG_HOUR, hour);
        args.putInt(ARG_MINUTE, minute);
        fragment.setArguments(args);

        return fragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        if (args != null) {
            int hour = args.getInt(ARG_HOUR);
            int minute = args.getInt(ARG_MINUTE);

            TimePickerDialog dialog = new TimePickerDialog(getActivity(), this, hour, minute, true);

            return dialog;
        } else {
            return null;
        }
    }

    @Override
    public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
        if (getTargetFragment() == null) {
            return;
        }

        Intent intent = new Intent();
        intent.putExtra(EXTRA_HOUR, hourOfDay);
        intent.putExtra(EXTRA_MINUTE, minute);

        getTargetFragment().onActivityResult(getTargetRequestCode(), Activity.RESULT_OK, intent);
    }
}

File: app/src/main/java/com/mariotti/developer/futureclock/ui/fragments/TimePickerFragment.kt
package com.mariotti.developer.futureclock.ui.fragments

import android.app.Activity
import android.app.Dialog
import android.app.TimePickerDialog
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.util.Log
import android.widget.TimePicker

class TimePickerFragment private constructor() : DialogFragment(), TimePickerDialog.OnTimeSetListener {

    companion object {
        val EXTRA_HOUR = "com.mariotti.developer.futureclock.controller.hour"
        val EXTRA_MINUTE = "com.mariotti.developer.futureclock.controller.minute"

        private val TAG = "TimePickerFragment"
        private val ARG_HOUR = "ARG_HOUR"
        private val ARG_MINUTE = "ARG_MINUTE"

        fun newInstance(hour: Int, minute: Int): TimePickerFragment {
            val fragment = TimePickerFragment()
            val args = Bundle()
            args.putInt(ARG_HOUR, hour % 24)
            args.putInt(ARG_MINUTE, minute % 60)
            fragment.arguments = args

            Log.d(TAG, "newInstance")

            return fragment
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val args = arguments
        val hour = args.getInt(ARG_HOUR)
        val minute = args.getInt(ARG_MINUTE)

        val dialog = TimePickerDialog(activity, this, hour, minute, true)

        return dialog
    }

    override fun onTimeSet(view: TimePicker, hourOfDay: Int, minute: Int) {
        targetFragment?.let {
            val intent = Intent()
            intent.putExtra(EXTRA_HOUR, hourOfDay)
            intent.putExtra(EXTRA_MINUTE, minute)

            it.onActivityResult(targetRequestCode, Activity.RESULT_OK, intent)
        }
    }
}
